### PR TITLE
Add API tests for index, stateManager

### PR DESCRIPTION
**Pull request by s1na**
This PR is part of the [series](https://github.com/ethereumjs/ethereumjs-vm/pull/315) for adding API test coverage. It:

- Adds tests for `index.js`
- Adds basic tests for `stateManager.js`

### CODE REVIEW

File: lib/fakeBlockChain.js [hunk 1]
```diff
@@ -2,6 +2,7 @@ const Buffer = require('safe-buffer').Buffer
 const utils = require('ethereumjs-util')
 
 module.exports = {
+  fake: true,
```
Comment (holgerd77) [inline lib/fakeBlockChain.js]: Ok.

File: lib/stateManager.js [hunk 2]
```diff
@@ -227,7 +227,7 @@ proto.revert = function (cb) {
 //
 proto.getStateRoot = function (cb) {
   var self = this
-  self.cacheFlush(function (err) {
+  self.cache.flush(function (err) {
```
Comment (holgerd77) [inline lib/stateManager.js]: Ok.

File: tests/api/index.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+const testData = require('./testdata.json')
+
+tape('VM with fake blockchain', (t) => {
+  t.test('should insantiate without params', (st) => {
+    const vm = new VM()
+    st.ok(vm.stateManager)
+    st.equal(vm.stateManager.trie.root, util.KECCAK256_RLP, 'it has default trie')
+    st.ok(vm.blockchain.fake, 'it has fake blockchain by default')
+    st.end()
+  })
```
Comment (holgerd77) [inline tests/api/index.js]: Ok. This is so cool, that this stuff get's explicit here. 👍

File: tests/api/index.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+  t.test('should be able to activate precompiles', (st) => {
+    let vm = new VM({ activatePrecompiles: true })
+    st.notEqual(vm.stateManager.trie.root, util.KECCAK256_RLP, 'it has different root')
+    st.end()
```
Comment (holgerd77) [inline tests/api/index.js]: Nice check, ok.

File: tests/api/index.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+  t.test('should only accept valid chain and fork', (st) => {
+    let vm = new VM({ chain: 'ropsten', hardfork: 'byzantium' })
+    st.equal(vm.stateManager._common.param('gasPrices', 'ecAdd'), 500)
+
+    try {
+      vm = new VM({ chain: 'mainchain', hardfork: 'homestead' })
+      st.fail('should have failed for invalid chain')
+    } catch (e) {
+      st.ok(e.message.includes('not supported'))
```
Comment (holgerd77) [inline tests/api/index.js]: Ok.

File: tests/api/index.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+    st.end()
+  })
+
+  t.test('should run blockchain without blocks', async (st) => {
+    const vm = new VM()
+    const run = promisify(vm.runBlockchain.bind(vm))
+    await run()
+    st.end()
+  })
```
Comment (holgerd77) [inline tests/api/index.js]: Ok.

File: tests/api/index.js [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+})
+
+tape('VM with blockchain', (t) => {
+  t.test('should instantiate', (st) => {
+    const vm = setupVM()
+    st.equal(vm.stateManager.trie.root, util.KECCAK256_RLP, 'it has default trie')
+    st.notOk(vm.stateManager.fake, 'it doesn\'t have fake blockchain')
+    st.end()
```
Comment (holgerd77) [inline tests/api/index.js]: Ok.

File: tests/api/index.js [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+  })
+
+  t.test('should run blockchain without blocks', async (st) => {
+    const vm = setupVM()
+    await runBlockchainP(vm)
+    st.end()
+  })
```
Comment (holgerd77) [inline tests/api/index.js]: Ok.

File: tests/api/index.js [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+      head.hash().toString('hex'),
+      testData.blocks[0].blockHeader.hash.slice(2)
+    )
+
+    vm.runBlock = (block, cb) => cb(new Error('test'))
+    runBlockchainP(vm)
+      .then(() => st.fail('it hasn\'t returned any errors'))
+      .catch((e) => {
+        st.equal(e.message, 'test', 'it has correctly propagated runBlock\'s error')
+        st.end()
```
Comment (holgerd77) [inline tests/api/index.js]: Ok.

File: tests/api/index.js [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+      testData.blocks[0].blockHeader.hash.slice(2)
+    )
+
+    const setupPreP = promisify(setupPreConditions)
+    await setupPreP(vm.stateManager.trie, testData)
+
+    await runBlockchainP(vm)
+
+    st.end()
+  })
```
Comment (holgerd77) [inline tests/api/index.js]: Ok.

File: tests/api/stateManager.js [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+    st.equal(stateManager.trie.root, util.KECCAK256_RLP, 'it has default root')
+    stateManager.getStateRoot((err, res) => {
+      st.error(err, 'getStateRoot returns no error')
+      st.equal(res, util.KECCAK256_RLP, 'it has default root')
+    })
+
+    st.equal(stateManager._common.hardfork(), 'byzantium', 'it has default hardfork')
+    st.end()
+  })
```
Comment (holgerd77) [inline tests/api/stateManager.js]: Ok.

File: tests/api/stateManager.js [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+      account
+    )
+
+    let res = await promisify(stateManager.getAccount.bind(stateManager))(
+      'a94f5374fce5edbc8e2a8697c15331677e6ebf0b'
+    )
+
+    st.equal(res.balance.toString('hex'), 'fff384')
+    st.end()
+  })
```
Comment (holgerd77) [inline tests/api/stateManager.js]: Ok.

**Comment by s1na** [comment]
Hey @holgerd77, thanks, I'm glad the quality was sufficient :)
Yes, this was the last PR I had in mind. If there's something missing, please let me know.

**Pull request merged**